py/server/deephaven/dbc.py

```diff
@@ -0,0 +1,48 @@
+#
+#     Copyright (c) 2016-2022 Deephaven Data Labs and Patent Pending
+#
+"""This module supports ingesting data from external databases into Deephaven via database connectivity interfaces.
+
+Currently, it requires that application must use Turbodbc (https://turbodbc.readthedocs.io/en/latest/index.html) to
+access a supported relational database. Turbodbc has built-in Apache Arrow support which enables Deephaven to
+achieve maximum performance and efficiency when ingesting relational data.
+
+Looking ahead, support of Arrow Database Connectivity (ADBC) https://arrow.apache.org/docs/dev/format/ADBC.html
+will be added once it reaches stable status. ADBC offers a standard API to fetch data in Arrow format from data
+sources that support Arrow natively as well as ODBC-only databases.
+"""
+from typing import Any
+
+from deephaven.table import Table
+from deephaven import arrow as dharrow
+from deephaven import DHError
+
+try:
+    import turbodbc
+except ImportError:
+    raise DHError(message="import turbodbc failed")
+
+
+def read_cursor(cursor: Any) -> Table:
+    """Converts the result set of the provided cursor into a Deephaven table.
+
+    Args:
+        cursor (Any): a Turbodbc cursor. Prior to it being passed in, its execute() method must be called to run a query
+            operation that produces a result set
+
+    Returns:
+        a new Table
+
+    Raises:
+        DHError
+    """
+
+    if not isinstance(cursor, turbodbc.cursor.Cursor):
+        raise DHError(message='')
+
+    try:
+        pa_table = cursor.fetchallarrow()
+    except Exception as e:
+        raise DHError(e, message="failed to fetch ODBC result as a Arrow table.") from e
+
+    return dharrow.to_table(pa_table)
```

Comment:
@rcaudy the 2nd case might be a useful feature but I am not sure if this is something that is truly part of an ingestion service for DHC.

Reply:
We have no code for writing parquet from data that's not in a DH table. This seems entirely out of scope for us at this time. Likely there is existing tooling based on `pyarrow`, and no need for us to duplicate any functionality here.

Reply:
1) If there is existing tooling, we need a link so that DevRel can document it.
2) If there is no existing tooling, we need a ticket to implement something.

Reply:
A quick search for 'export relational data to Parquet files' or 'export Postgres tables to Parquet files' gets many hits that seem legit. So there are some tools or built-in database features that give users the ability to do that.

Reply:
> 1. If there is existing tooling, we need a link so that DevRel can document it.
> 2. If there is no existing tooling, we need a ticket to implement something.

Why? Pulling data out of unrelated tools and writing it into formats we don't own isn't really "on brand" for us.

Reply:
What *would* be good is a streaming ingest to Deephaven's incremental persistence format, when/if that is brought over from enterprise.

Reply:
"on brand" is making data useful.  For anyone doing data science, this is a common use case they will run into.

Reply:
Documentation ticket:  https://github.com/deephaven/deephaven.io/issues/2121